buku

```diff
@@ -921,98 +1013,192 @@ class BukuDb:
-            if bad:
+        fetch_title = {url, title_in, tags_in, desc, immutable} == {None}
+        network_test = url_redirect or tag_redirect or tag_error or del_error or export_on or fetch_title
+        if url and title_in is None:
+            network_test = False
+            _url = url or self.get_rec_by_id(index).url
+            result = fetch_data(_url)
+            if result.bad:
                 print('Malformed URL')
-            elif mime:
+            elif result.mime:
                 LOGDBG('HTTP HEAD requested')
-            elif title_to_insert == '':
+            elif not result.title:
                 print('No title')
             else:
-                LOGDBG('Title: [%s]', title_to_insert)
+                LOGDBG('Title: [%s]', result.title)
 
-            if not desc:
-                if not pdesc:
-                    pdesc = ''
+            if result.desc and not desc:
                 query += ' desc = ?,'
-                arguments += (pdesc,)
-                to_update = True
-        elif not to_update and not tag_modified:
-            ret = self.refreshdb(index, threads)
-            if ret and index and self.chatty:
-                self.print_rec(index)
-            return ret
+                arguments += (result.desc,)
+
+            if url_redirect and result.url != _url:
+                url = result.url
 
-        if title_to_insert is not None:
+            if result.fetch_status in export_on:  # storing the old URL
+                self._to_export[url or _url] = _url
+        else:
+            result = FetchResult(url, title_in)
+
+        if result.title is not None:
             query += ' metadata = ?,'
-            arguments += (title_to_insert,)
-            to_update = True
+            arguments += (result.title,)
 
-        if not to_update:  # Nothing to update
-            # Show bookmark if tags were appended to deleted
-            if tag_modified and self.chatty:
-                self.print_rec(index)
-            return ret
+        # Update URL if passed as argument
+        if url:
+            query += ' URL = ?,'
+            arguments += (url,)
+
+        if result.fetch_status in (del_error or []):
+            if result.fetch_status in export_on:  # storing the old record
+                self._to_export[url] = self.get_rec_by_id(index)
+            LOGERR('HTTP error %s', result.fetch_status)
+            return self.delete_rec(index)
 
-        if index == 0:  # Update all records
+        if not indices and (arguments or tags_in):
             resp = read_in('Update ALL bookmarks? (y/n): ')
             if resp != 'y':
                 return False
 
-            query = query[:-1]
-        else:
-            query = query[:-1] + ' WHERE id = ?'
-            arguments += (index,)
+        if network_test:  # doing this before updates to backup records to-be-deleted in their original state
+            custom_tags = (tags_in if (tags_in or '').startswith(DELIM) else None)
```

Comment:
`custom_tags` is used _only_ when tags _overwrite_ was requested, and ensures that redirect/error tags are still applied.